File: b52-bomber/feature-tik/src/main/java/com/mobile/app/bomber/tik/home/items/CommentItem.java
package com.mobile.app.bomber.tik.home.items;

import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.util.Preconditions;
import com.mobile.app.bomber.data.http.entities.ApiComment;
import com.mobile.app.bomber.data.http.entities.ApiVideo;
import com.mobile.app.bomber.data.repository.MappersKt;
import com.mobile.app.bomber.runner.base.PrefsManager;
import com.mobile.app.bomber.tik.R;
import com.mobile.app.bomber.tik.databinding.ItemCommentABinding;
import com.mobile.app.bomber.tik.databinding.ItemCommentBBinding;
import com.mobile.app.bomber.tik.databinding.ItemCommentCBinding;
import com.mobile.guava.android.mvvm.AndroidX;
import com.mobile.guava.android.mvvm.Msg;
import com.mobile.guava.android.ui.screen.ScreenUtilsKt;
import com.mobile.guava.android.ui.view.text.MySpannable;
import com.mobile.guava.jvm.date.Java8TimeKt;
import com.pacific.adapter.AdapterViewHolder;
import com.pacific.adapter.SimpleRecyclerItem;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public abstract class CommentItem extends SimpleRecyclerItem {

    public static final int SLOT_COUNT = 3;

    @NonNull
    public final ApiComment.Comment data;
    public final ApiVideo.Video video;

    public final int type;
    public final boolean isMe;
    protected MySpannable comment;

    public CommentItem(@NonNull ApiComment.Comment data, int type, ApiVideo.Video video) {
        this.type = type;
        this.data = data;
        this.video = video;
        if (type == 3) {
            this.isMe = false;
        } else {
            this.isMe = video.getOwner() == data.getUid();
        }
    }

    protected MySpannable buildComment() {
        if (comment == null) {
            int atColor = ContextCompat.getColor(AndroidX.INSTANCE.myApp(), R.color.comment_at);
            int color = ContextCompat.getColor(AndroidX.INSTANCE.myApp(), R.color.comment_name);
            String replay = data.getReplayText();

            String ago = Java8TimeKt.ago(data.getTime() * 1000L, System.currentTimeMillis());
            if (ago.equals("0秒前")){
                ago = "1秒前";
            }
            comment = new MySpannable(replay + data.getContent() + "\u3000" + ago)
                    .findAndSpan(replay, () -> new ForegroundColorSpan(atColor))
                    .findAndSpan(ago, () -> new ForegroundColorSpan(color));
            data.getAtTexts().stream().forEach(it -> comment.findAndSpan(
                    it,
                    () -> new ForegroundColorSpan(atColor))
            );
        }
        return comment;
    }

    public static class TypeA extends CommentItem {

        public TypeA(@NonNull ApiComment.Comment data, ApiVideo.Video video) {
            super(data, 1,video);
        }

        @Override
        public void bind(@NotNull AdapterViewHolder holder) {
            ItemCommentABinding binding = holder.binding(ItemCommentABinding::bind);
            binding.txtName.setText(data.getUsername());
            if (isMe) {
                binding.txtLabel.setVisibility(View.VISIBLE);
            } else {
                binding.txtLabel.setVisibility(View.INVISIBLE);
            }
            if (TextUtils.isEmpty(data.getContent())) {
                binding.txtComment.setText("");
            } else {
                binding.txtComment.setText(buildComment());
            }
            binding.txtLike.setSelected(data.isLiking());
            binding.txtLike.setText(String.valueOf(data.getLikeCount()));
            holder.attachImageLoader(R.id.img_profile);
            holder.attachOnClickListener(R.id.txt_like);
            holder.attachOnClickListener(R.id.item_comment_a);
            holder.attachOnLongClickListener(R.id.item_comment_a);

            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) binding.getRoot().getLayoutParams();
            params.width = ScreenUtilsKt.getScreen().x;
            binding.getRoot().setLayoutParams(params);
        }

        @Override
        public void bindPayloads(@NotNull AdapterViewHolder holder, @Nullable List<?> payloads) {
            ItemCommentABinding binding = holder.binding(ItemCommentABinding::bind);
            binding.txtLike.setSelected(data.isLiking());
            binding.txtLike.setText(String.valueOf(data.getLikeCount()));
        }

        @Override
        public int getLayout() {
            return R.layout.item_comment_a;
        }
    }

    public static class TypeB extends CommentItem {

        public TypeB(@NonNull ApiComment.Comment data, ApiVideo.Video video) {
            super(data, 2,video);
        }

        @Override
        public void bind(@NotNull AdapterViewHolder holder) {
            ItemCommentBBinding binding = holder.binding(ItemCommentBBinding::bind);
            binding.txtName.setText(data.getUsername());
            if (isMe) {
                binding.txtLabel.setVisibility(View.VISIBLE);
            } else {
                binding.txtLabel.setVisibility(View.INVISIBLE);
            }
            if (TextUtils.isEmpty(data.getContent())) {
                binding.txtComment.setText("");
            } else {
                binding.txtComment.setText(buildComment());
            }
            binding.txtLike.setSelected(data.isLiking());
            binding.txtLike.setText(String.valueOf(data.getLikeCount()));
            holder.attachImageLoader(R.id.img_profile);
            holder.attachOnClickListener(R.id.txt_like);
            holder.attachOnClickListener(R.id.item_comment_b);
            holder.attachOnLongClickListener(R.id.item_comment_b);

            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) binding.getRoot().getLayoutParams();
            params.width = ScreenUtilsKt.getScreen().x;
            binding.getRoot().setLayoutParams(params);
        }

        @Override
        public void bindPayloads(@NotNull AdapterViewHolder holder, @Nullable List<?> payloads) {
            ItemCommentBBinding binding = holder.binding(ItemCommentBBinding::bind);
            binding.txtLike.setSelected(data.isLiking());
            binding.txtLike.setText(String.valueOf(data.getLikeCount()));
        }

        @Override
        public int getLayout() {
            return R.layout.item_comment_b;
        }
    }

    public static class TypeC extends CommentItem {

        private final List<ApiComment.Comment> list;
        private int hidingCount;
        private int showingCount;
        private int lastShowingPager;

        public TypeC(@NonNull ApiComment.Comment data,ApiVideo.Video video) {
            super(data, 3,video);
            list = data.getChildren().subList(SLOT_COUNT, data.getChildren().size());
            Preconditions.checkArgument(list.size() > 0, "");
            hidingCount = list.size();
            showingCount = 0;
            lastShowingPager = 0;
        }

        @Override
        public void bind(@NotNull AdapterViewHolder holder) {
            ItemCommentCBinding binding = holder.binding(ItemCommentCBinding::bind);
            setComment(binding);
            holder.attachOnClickListener(R.id.item_comment_c);
        }

        @Override
        public void bindPayloads(@NotNull AdapterViewHolder holder, @Nullable List<?> payloads) {
            ItemCommentCBinding binding = holder.binding(ItemCommentCBinding::bind);
            setComment(binding);
        }

        @Override
        public int getLayout() {
            return R.layout.item_comment_c;
        }

        private void setComment(ItemCommentCBinding binding) {
            String comment;
            if (list.size() == hidingCount) {
                comment = "展开" + list.size() + "条回复";
            } else if (hidingCount == 0) {
                comment = "收起";
            } else {
                comment = "展开更多回复";
            }
            binding.txtComment.setText(comment);
        }

        public List<ApiComment.Comment> next() {
            List<ApiComment.Comment> next = MappersKt.forPage(list, lastShowingPager + 1, SLOT_COUNT);
            if (!next.isEmpty()) {
                lastShowingPager++;
                showingCount = showingCount + next.size();
                hidingCount = hidingCount - next.size();
            }
            return next;
        }

        public void reset() {
            hidingCount = list.size();
            showingCount = 0;
            lastShowingPager = 0;
        }

        public int size() {
            return list.size();
        }
    }


    public static class TypeD extends SimpleRecyclerItem {

        public TypeD() {
            super();
        }

        @Override
        public void bind(@NotNull AdapterViewHolder adapterViewHolder) {
        }

        @Override
        public int getLayout() {
            return R.layout.item_comment_d;
        }
    }
}

File: b52-bomber/feature-movie/src/main/java/com/mobile/app/bomber/movie/player/items/CommentItem.kt
package com.mobile.app.bomber.movie.player.items

import android.text.style.ForegroundColorSpan
import android.view.View
import com.mobile.app.bomber.data.http.entities.ApiComment
import com.mobile.app.bomber.data.repository.forPage
import com.mobile.app.bomber.movie.R
import com.mobile.app.bomber.movie.databinding.MovieItemCommentABinding
import com.mobile.app.bomber.movie.databinding.MovieItemCommentBBinding
import com.mobile.app.bomber.movie.databinding.MovieItemCommentCBinding
import com.mobile.guava.android.mvvm.AndroidX
import com.mobile.app.bomber.runner.base.PrefsManager
import com.mobile.guava.android.context.toColor
import com.mobile.guava.android.ui.view.text.MySpannable
import com.mobile.guava.jvm.date.ago
import com.pacific.adapter.AdapterViewHolder
import com.pacific.adapter.SimpleRecyclerItem

abstract class CommentItem(val data: ApiComment.Comment, val type: Int) : SimpleRecyclerItem() {

    protected val isMe: Boolean = if (type == 3) false else PrefsManager.getUserId() == data.uid
    protected val spannableComment: MySpannable by lazy {
        val userColor = AndroidX.myApp.toColor(R.color.movie_player_comment_user)
        val commentColor = AndroidX.myApp.toColor(R.color.movie_player_game_ad_desc)
        val ago = ago(data.time, System.currentTimeMillis())
        val replay = data.replayText

        return@lazy MySpannable(replay + data.content + "\u3000" + ago)
                .findAndSpan(replay) { ForegroundColorSpan(userColor) }
                .findAndSpan(ago) { ForegroundColorSpan(commentColor) }
                .apply {
                    data.atTexts.forEach {
                        findAndSpan(it) { ForegroundColorSpan(userColor) }
                    }
                }
    }

    class TypeA(data: ApiComment.Comment) : CommentItem(data, 1) {

        override fun bind(holder: AdapterViewHolder) {
            val binding: MovieItemCommentABinding = holder.binding(MovieItemCommentABinding::bind)
            binding.txtName.text = data.username
            if (isMe) {
                binding.txtLabel.visibility = View.VISIBLE
            } else {
                binding.txtLabel.visibility = View.INVISIBLE
            }
            binding.txtComment.text = spannableComment
            binding.txtLike.isSelected = data.isLiking
            binding.txtLike.text = data.likeCount.toString()

            holder.attachImageLoader(R.id.img_profile)
            holder.attachOnClickListener(R.id.txt_like)
            holder.attachOnClickListener(R.id.item_comment_a)
            holder.attachOnLongClickListener(R.id.item_comment_a)
        }

        override fun bindPayloads(holder: AdapterViewHolder, payloads: List<Any>?) {
            val binding: MovieItemCommentABinding = holder.binding()
            binding.txtLike.isSelected = data.isLiking
            binding.txtLike.text = data.likeCount.toString()
        }

        override fun getLayout(): Int = R.layout.movie_item_comment_a
    }

    class TypeB(data: ApiComment.Comment) : CommentItem(data, 2) {
        override fun bind(holder: AdapterViewHolder) {
            val binding: MovieItemCommentBBinding = holder.binding(MovieItemCommentBBinding::bind)
            binding.txtName.text = data.username
            if (isMe) {
                binding.txtLabel.visibility = View.VISIBLE
            } else {
                binding.txtLabel.visibility = View.INVISIBLE
            }
            binding.txtComment.text = spannableComment
            binding.txtLike.isSelected = data.isLiking
            binding.txtLike.text = data.likeCount.toString()
            holder.attachImageLoader(R.id.img_profile)
            holder.attachOnClickListener(R.id.txt_like)
            holder.attachOnClickListener(R.id.item_comment_b)
            holder.attachOnLongClickListener(R.id.item_comment_b)
        }

        override fun bindPayloads(holder: AdapterViewHolder, payloads: List<Any>?) {
            val binding: MovieItemCommentBBinding = holder.binding()
            binding.txtLike.isSelected = data.isLiking
            binding.txtLike.text = data.likeCount.toString()
        }

        override fun getLayout(): Int = R.layout.movie_item_comment_b
    }

    class TypeC(data: ApiComment.Comment) : CommentItem(data, 3) {

        private val list: List<ApiComment.Comment>
        private var hidingCount = 0
        private var showingCount = 0
        private var lastShowingPage = 0

        init {
            require(!data.children.isNullOrEmpty())
            list = data.children!!.subList(SLOT_COUNT, data.children!!.size)
            hidingCount = list.size
            showingCount = 0
            lastShowingPage = 0
        }

        override fun bind(holder: AdapterViewHolder) {
            val binding: MovieItemCommentCBinding = holder.binding(MovieItemCommentCBinding::bind)
            setComment(binding)
            holder.attachOnClickListener(R.id.item_comment_c)
        }

        override fun bindPayloads(holder: AdapterViewHolder, payloads: List<Any>?) {
            val binding: MovieItemCommentCBinding = holder.binding()
            setComment(binding)
        }

        override fun getLayout(): Int = R.layout.movie_item_comment_c

        private fun setComment(binding: MovieItemCommentCBinding) {
            val comment: String = when {
                list.size === hidingCount -> "展开" + list.size.toString() + "条回复"
                hidingCount === 0 -> "收起"
                else -> "展开更多回复"
            }
            binding.txtComment.text = comment
        }

        fun next(): List<ApiComment.Comment> {
            val next = list.forPage(lastShowingPage + 1, SLOT_COUNT)
            if (next.isNotEmpty()) {
                lastShowingPage++
                showingCount += next.size
                hidingCount -= next.size
            }
            return next
        }

        fun size(): Int = list.size

        fun reset() {
            hidingCount = list.size
            showingCount = 0
            lastShowingPage = 0
        }
    }

    class TypeD : SimpleRecyclerItem() {

        override fun bind(adapterViewHolder: AdapterViewHolder) {}

        override fun getLayout(): Int = R.layout.movie_item_comment_d
    }

    companion object {

        @get:JvmName("SLOT_COUNT")
        const val SLOT_COUNT = 3
    }
}